src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1070 @@
+    // because lo[first_violation] == 1. We can also see that any larger value
+    // of i would violate 3.1 since lo[first_violation] does not satisfy bits.
+    // As a result, we should find the last i upto first_violation such that
+    // lo[i] == zeros[i] == 0.
+    //
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 0 0 0 1 1 1 0
+    //   zeros = 0 0 0 1 0 1 0 0
+    //    ones = 1 0 0 0 0 0 1 1
+    //   1-vio = 0 0 0 0 0 0 0 1
+    //   0-vio = 0 0 0 0 0 1 0 0
+    // The first violation is the 6th bit, which should be 0. We want to flip
+    // it to 0. However, since we must obtain a value larger than lo, we must
+    // find an earlier bit that can be flipped from 0 to 1. The 5th cannot be
+    // the bit we are looking for, because it is already 1, the 4th bit also
+    // cannot be, because it must be 0. As a result, the last bit we can flip,
+    // which is the first different bit between the result and lo must be the
+    // 3rd bit. As a result, the result must not be smaller than:
+    //           1 0 1 0 0 0 0 0
+    // This one satisfies zeros so we can use the logic in the previous case,
+    // just OR with ones to obtain the final result, which is:
+    //           1 0 1 0 0 0 1 1
+
+    juint first_violation = count_leading_zeros(zero_violation);
+    // This masks out all bits after the first violation
+    //           1 1 1 1 1 0 0 0
+    U find_mask = ~(std::numeric_limits<U>::max() >> first_violation);
+    // A bit that is 0 in both lo and zeros must be 0 in either
+    //           1 0 0 1 1 1 1 0
+    U either = lo | bits._zeros;
+    // This is all the bits that are not lower than first_violation and are 0
+    // in both lo and zeros. Note that there may not exist such bits. In those
+    // cases tmp == 0 and there is no value not less than lo and satisfies
+    // bits. We will expand the implication of such cases below.
+    //           0 1 1 0 0 0 0 0
+    U tmp = ~either & find_mask;
+    // According to the conclusion in the overview of this case above, i is the
+    // last bit being 0 in both lo and zeros that is upto the first violation,
+    // which is the last set bit of tmp.
+    // In our example, i == 2 here, we want to obtain the value with only the
+    // bit i set, this is equivalent to extracting the last set bit of tmp, do
+    // it directly without going through i.
+    //           0 0 1 0 0 0 0 0
+    U alignment = tmp & (-tmp);
+    // Set the bit of lo at i and unset all the bits after, this is the smallest
+    // value that satisfies bits._zeros. Similar to the above case, this is
+    // similar to aligning lo upto alignment. Also similar to the above case,
+    // this computation cannot overflow.
+    // We now have:
+    // - new_lo[x] = lo[x], for 0 <= x < i (2.5)
+    // - new_lo[i] = 1                     (2.6)
+    // - new_lo[x] = 0, for x > i          (not yet 2.7)
+    //           1 0 1 0 0 0 0 0
```

Comment:
That's a little misleading if there is no such `i`...
In that case `alignment = 0`, and `-alignment = 0`, and so `new_lo = 0`... which is not exactly an overflow, but smaller than `lo`, so kinda an overflow ;)

Reply:
```suggestion
    // similar to aligning lo upto alignment. Also similar to the above case,
    // this computation cannot overflow.
    // We now have:
    // - new_lo[x] = lo[x], for 0 <= x < i (2.5)
    // - new_lo[i] = 1                     (2.6)
    // - new_lo[x] = 0, for x > i          (not yet 2.7)
    // If there is no such candidate, and no such i, then new_lo = 0.
    //           1 0 1 0 0 0 0 0
```